src/zenml/integrations/modal/step_operators/modal_step_operator.py

```diff
@@ -0,0 +1,215 @@
+                    step_name=step_name,
+                )
+                builds.append(build)
+
+        return builds
+
+    def launch(
+        self,
+        info: "StepRunInfo",
+        entrypoint_command: List[str],
+        environment: Dict[str, str],
+    ) -> None:
+        """Launch a step run on Modal.
+
+        Args:
+            info: The step run information.
+            entrypoint_command: The entrypoint command for the step.
+            environment: The environment variables for the step.
+
+        Raises:
+            subprocess.CalledProcessError: If the modal app exits with a non-zero code.
+            ValueError: If no Docker credentials are found for the container registry.
+        """
+        settings = cast(ModalStepOperatorSettings, self.get_settings(info))
+        image_name = info.get_image(key=MODAL_STEP_OPERATOR_DOCKER_IMAGE_KEY)
+        zc = Client()
+        stack = zc.active_stack
+
+        if not stack.container_registry:
+            raise ValueError(
+                "No Container registry found in the stack. "
+                "Please add a container registry and ensure "
+                "it is correctly configured."
+            )
+
+        if docker_creds := stack.container_registry.credentials:
+            docker_username, docker_password = docker_creds
+        else:
+            raise ValueError(
+                "No Docker credentials found for the container registry."
+            )
+
+        # get the pydantic version in local environment
+        # use it to install the correct version of pydantic in the modal app
+        # since it overwrites with 1.x on top of our image
+        pydantic_version = importlib.metadata.version("pydantic")
+        major, minor, *_ = pydantic_version.split(".")
+
+        # Construct the decorator arguments based on the settings
+        decorator_args = []
+        if settings.region is not None:
+            decorator_args.append(f"region='{settings.region}'")
+        if settings.cloud is not None:
+            decorator_args.append(f"cloud='{settings.cloud}'")
+
+        # if resource settings are set, add them to the decorator args
+        resource_settings = info.config.resource_settings
+        if settings.gpu is not None:
+            gpu_str = settings.gpu
+            if resource_settings.gpu_count is not None:
+                gpu_str += f":{resource_settings.gpu_count}"
+            decorator_args.append(f"gpu='{gpu_str}'")
+        if resource_settings.cpu_count is not None:
+            decorator_args.append(f"cpu={resource_settings.cpu_count}")
+        if resource_settings.memory is not None:
+            decorator_args.append(
+                f"memory={resource_settings.get_memory(ByteUnit.MB)}"
+            )
+
+        decorator_args_str = ", ".join(decorator_args)
+
+        with tempfile.NamedTemporaryFile(
+            mode="w", delete=False, suffix=".py"
+        ) as tmp:
+            modal_code = f"""import modal
+
```

Comment:
```suggestion

        import modal

        my_secret = modal.Secret.from_dict(
            {
                "REGISTRY_USERNAME": docker_username,
                "REGISTRY_PASSWORD": docker_password,
            }
        )

        zenml_image = (
            modal.Image.from_registry(tag=image_name, secret=my_secret)
            .env(environment)
            .pip_install(f"pydantic~={major}.{minor}")
            .run_commands(" ".join(entrypoint_command))
        )

        resource_settings = info.config.resource_settings

        sb = modal.Sandbox(
            "bash",
            "-c",
            cmd, # TODO : Add cmd
            image=zenml_image,
            gpu=f":{resource_settings.gpu_count}",
            cpu=resource_settings.cpu_count,
            memory=resource_settings.get_memory(ByteUnit.MB),
            cloud=settings.cloud,
            region=settings.region,
        )

        sb.wait()

        stdout = sb.stdout()
        stderr = sb.stderr()
        returncode = sb.returncode()

        return stdout, stderr, returncode

```

Reply:
Could the rest of this function be replaced with a modal sandbox?

Reply:
Or actually using ephemeral apps would probably be better than sandbox in your case:

```py
        import modal

        my_secret = modal.Secret.from_dict(
            {
                "REGISTRY_USERNAME": docker_username,
                "REGISTRY_PASSWORD": docker_password,
            }
        )

        zenml_image = (
            modal.Image.from_registry(tag=image_name, secret=my_secret)
            .env(environment)
            .pip_install(f"pydantic~={major}.{minor}")
            .run_commands(" ".join(entrypoint_command))
        )

        resource_settings = info.config.resource_settings

        app = modal.App(
            "example-pandas",
            image=zenml_image,
            gpu=f":{resource_settings.gpu_count}",
            cpu=resource_settings.cpu_count,
            memory=resource_settings.get_memory(ByteUnit.MB),
            cloud=settings.cloud,
            region=settings.region,
        )

        with app.run():
            # run your step code here!
            # run_step()
            pass

```

Reply:
Sandbox has the advantage of not installing a modal layer of requirements on top of the image, but the disadvantage of not having streaming logs or any observability into what's going on in the image. I couldn't figure out how to get the logs in the dashboard, even. Is there a way?